ext/drpttracer/drpttracer.c

```diff
@@ -0,0 +1,579 @@
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* DynamoRIO Intel PT Tracing Extension. */
+
+#include <errno.h>
+#include <stdint.h>
+#include <linux/perf_event.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/ioctl.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+#include "../../core/unix/include/syscall_linux_x86.h" // for SYS_perf_event_open
+#include "dr_api.h"
+#include "drpttracer.h"
+
+#if !defined(X86_64) && !defined(LINUX)
+#    error "This is only for Linux x86_64."
+#endif
+
+#ifdef DEBUG
+#    define ASSERT(x, msg) DR_ASSERT_MSG(x, msg)
+#else
+#    define ASSERT(x, msg) /* nothing */
+#endif
+
+#ifdef DEBUG
+#    define ERRMSG(...)                      \
+        do {                                 \
+            dr_fprintf(STDERR, __VA_ARGS__); \
+        } while (0)
+#else
+#    define ERRMSG(...) /* nothing */
+#endif
+
+/* drpttracer supports multiple type's tracing. However, perf_event_attr's initialization
+ * causes significant overhead. So drpttracer stores all tracing types' perf_event_attr as
+ * global static variables, then drpttracer can directly use them when the caller want to
+ * start tracing.
+ */
+static struct perf_event_attr user_only_pe_attr;
+static struct perf_event_attr kernel_only_pe_attr;
+static struct perf_event_attr user_kernel_pe_attr;
+
+/* pt_shared_metadata is used to store shared meta data: cpu_family, cpu_model,
+ * cpu_stepping.
+ */
+static pt_metadata_t pt_shared_metadata;
+
+/***************************************************************************
+ * UTTILITY FUNCTIONS
+ */
+
+static bool
+read_file_to_buf(IN const char *filename, OUT char **buf, OUT uint64_t *buf_size)
+{
+    char *local_buf = NULL;
+    uint64_t local_buf_size = 0;
+    file_t f = dr_open_file(filename, DR_FILE_READ);
+    if (f == INVALID_FILE) {
+        ERRMSG("failed to open file %s\n", filename);
+        return false;
+    }
+
+    if (!dr_file_size(f, &local_buf_size)) {
+        ERRMSG("failed to get file size of %s\n", filename);
+        dr_close_file(f);
+        return false;
+    }
+
+    local_buf = (char *)dr_global_alloc(local_buf_size);
```

Comment:
Can we use thread_alloc instead?

Reply:
This function will be called before all thread start events happen. Could we use the thread_alloc at that time?